### PR TITLE
Add whenValue to await a value that matches certain conditions

### PR DESCRIPTION
**Pull request by reuzel**
Hereby a proposal for an addition that allows the creation of a promise that resolves when a certain observed value become available and matches an (optionally) provided predicate. This allows the creation of a promise that resolves when a certain value is added to a map, when a certain value matches some condition, etc. The difference with mobx-when is that the promise resolves with the actual observed value.

Regards,
Joost

### CODE REVIEW

File: src/whenValue.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+ *  *
+ * @param generator - selects the value to be observed
+ * @param predicate - optional predicate to indicate when the observed value should resolve the promise. By default accepts all values.
+ * @returns a Promise that resolves with the observed value when it fulfills the conditions of the predicate
+ */
+export function whenValue<T>(
+    generator: () => T | undefined,
+    predicate: (value: T) => boolean = () => true
+): Promise<T> & { cancel: () => void } {
+    let cancel: () => void = () => undefined
```
Comment (NaridaL) [inline src/whenValue.ts:30]: You can do `let cancel!: () => void` to avoid assigning a value which will never be used.

File: src/whenValue.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        const disposer = autorun((reaction) => {
+            try {
+                const value = computed(generator).get()
+                if (value !== undefined && predicate(value)) {
+                    resolve(value)
+                    reaction.dispose()
+                }
+            } catch (err) {
+                reject(err)
+                reaction.dispose()
```
Comment (NaridaL) [inline src/whenValue.ts:42]: reject can't throw, right?

File: src/whenValue.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+ * The operation can be cancelled by the cancel operation added to the returned promise. This will reject the Promise.
+ *
+ * @example
+ * const store = observed({
+ *   foo:""
+ * });
+ *
+ * //wait until store.foo contains the character "b"
+ * await whenValue(()=>store.foo, (value)=>value.indexOf("b")>=0);
+ *  *
```
Comment (NaridaL) [inline src/whenValue.ts:21]: Can we format the example code like the actual code?

File: src/whenValue.ts [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+        cancel = () => {
+            disposer()
+            reject("WHEN_VALUE_CANCELLED")
+        }
+    })
+
+    const res = p as any
+    res.cancel = cancel
+    return res
```
Comment (NaridaL) [inline src/whenValue.ts:54]: You can do `Object.assign(new Promise(...), { cancel })` to avoid casting.

**Review by NaridaL — CHANGES_REQUESTED** [review]
You need to update the readme by running `yarn build-docs`.

**Comment by mweststrate** [comment]
Closing for inactivity.

**Pull request closed without merging**